gmeta/codegen/src/lib.rs

```diff
@@ -1,518 +1,503 @@
-                funcs.insert(#name.to_string(), repr);
+            first_type => {
+                if first_type != state_type {
+                    error!(first_type, "first argument's type must be `State`");
+                }
             }
         }
-    };
-
-    let mut block = quote! {
-        use gmeta::*;
 
-        let mut funcs: BTreeMap<String, TypesRepr> = Default::default();
-        let mut registry = Registry::new();
-    };
+        // Checking the rest of arguments
 
-    for (name, io) in map {
-        block.extend(add_fn(name, io));
-    }
+        let mut arguments = vec![];
 
-    block.extend(quote! {
-        let metawasm_data = MetawasmData {
-            funcs, registry: PortableRegistry::from(registry).encode(),
-        };
-        gstd::msg::reply(metawasm_data, 0).expect("Failed to share metadata");
-    });
+        for argument in inputs {
+            if let FnArg::Typed(argument) = argument {
+                if let Err(error) = has_attributes(&argument.attrs, "mustn't have attributes") {
+                    return error;
+                }
 
-    let res = quote! {
-        #[no_mangle]
-        extern "C" fn metadata() {
-            #block
+                arguments.push((argument.pat, argument.ty));
+            } else {
+                // The rest of arguments can't be the `self` argument because
+                // the compiler won't allow this.
+                unreachable!("unexpected `self` argument");
+            }
         }
-    };
 
-    res.into()
-}
+        // Checking an output
 
-fn construct_abi(funcs: Vec<ItemFn>) -> TokenStream {
-    let mut res = proc_macro2::TokenStream::new();
-    for mut func in funcs {
-        let span = func.span();
-        let prev_inputs = func.sig.inputs.clone();
-        let prev_ident = func.sig.ident;
-        let new_ident = syn::Ident::new(&format!("_{prev_ident}"), prev_ident.span());
-        func.sig.ident = new_ident.clone();
-
-        func.to_tokens(&mut res);
-
-        func.attrs = vec![Attribute {
-            pound_token: syn::token::Pound(span),
-            style: syn::AttrStyle::Outer,
-            bracket_token: syn::token::Bracket(span),
-            path: syn::Path {
-                leading_colon: None,
-                segments: Default::default(),
-            },
-            tokens: quote!(no_mangle),
-        }];
-
-        let mut sig = func.sig;
-        sig.ident = prev_ident;
-        sig.abi = Some(syn::Abi {
-            extern_token: syn::token::Extern(span),
-            name: Some(LitStr::new("C", span)),
-        });
-        sig.inputs = Default::default();
-        sig.output = ReturnType::Default;
-        func.sig = sig;
+        let return_type = match signature.output {
+            ReturnType::Default => error!(signature_span, "return type must be specified"),
+            ReturnType::Type(_, return_type) => {
+                if let Type::Tuple(tuple) = return_type.as_ref() {
```

Comment:
recheck please, that syn has no other ways to detect `unit` type

Reply:
https://github.com/dtolnay/syn/pull/858#pullrequestreview-437074058
No, it doesn't.